File: ios/Packages/GodstoneFoundation/Sources/GodstoneMesh/BleRecord.swift
import Foundation
import CryptoKit

/// ADR-002 Canonical BLE Record Layer Types & Constants (Phase C8.4C).
public enum BleRecordType: UInt8, Sendable, CaseIterable, Equatable {
    case hs1 = 0x11
    case hs2 = 0x12
    case hs3 = 0x14
    case data = 0x18
    case close = 0x21
}

public enum BleRecordConstants {
    public static let magic: UInt8 = 0x47
    public static let headerBytes: Int = 8
    public static let maxRecord: Int = 16384
    public static let maxFragments: Int = 64
    public static let maxConcurrent: Int = 4
    public static let reassemblyTimeoutSeconds: TimeInterval = 30.0
}

public struct BleRecordHeader: Equatable, Sendable {
    public let magic: UInt8
    public let recordType: BleRecordType
    public let recordSeq: UInt8
    public let fragIndex: UInt8
    public let fragCount: UInt8
    public let totalLen: UInt16
    public let headerCheck: UInt8

    public init(
        magic: UInt8 = BleRecordConstants.magic,
        recordType: BleRecordType,
        recordSeq: UInt8,
        fragIndex: UInt8,
        fragCount: UInt8,
        totalLen: UInt16,
        headerCheck: UInt8
    ) {
        self.magic = magic
        self.recordType = recordType
        self.recordSeq = recordSeq
        self.fragIndex = fragIndex
        self.fragCount = fragCount
        self.totalLen = totalLen
        self.headerCheck = headerCheck
    }
}

public struct BleRecordFragment: Equatable, Sendable {
    public let header: BleRecordHeader
    public let payload: Data

    public init(header: BleRecordHeader, payload: Data) {
        self.header = header
        self.payload = Data(payload)
    }
}

public struct BleReassembledRecord: Equatable, Sendable {
    public let recordType: BleRecordType
    public let recordSeq: UInt8
    public let payload: Data

    public init(recordType: BleRecordType, recordSeq: UInt8, payload: Data) {
        self.recordType = recordType
        self.recordSeq = recordSeq
        self.payload = Data(payload)
    }
}

public enum BleRecordCodec {

    public static func computeHeaderCheck(
        _ b0: UInt8,
        _ b1: UInt8,
        _ b2: UInt8,
        _ b3: UInt8,
        _ b4: UInt8,
        _ b5: UInt8,
        _ b6: UInt8
    ) -> UInt8 {
        b0 ^ b1 ^ b2 ^ b3 ^ b4 ^ b5 ^ b6
    }

    public static func computeHeaderCheck(_ data7: Data) -> UInt8 {
        let bytes = [UInt8](data7)
        precondition(bytes.count == 7, "computeHeaderCheck requires exactly 7 bytes")
        return computeHeaderCheck(
            bytes[0], bytes[1], bytes[2], bytes[3], bytes[4], bytes[5], bytes[6]
        )
    }

    public static func canonicalFragmentBounds(
        totalLen: Int,
        fragCount: Int,
        fragIndex: Int
    ) throws -> (start: Int, end: Int) {
        if totalLen == 0 {
            guard fragCount == 1, fragIndex == 0 else {
                throw NSError(domain: "BleRecordCodec", code: 1, userInfo: [NSLocalizedDescriptionKey: "totalLen=0 requires fragCount=1, fragIndex=0"])
            }
            return (0, 0)
        }

        guard fragCount >= 1, fragCount <= BleRecordConstants.maxFragments else {
            throw NSError(domain: "BleRecordCodec", code: 2, userInfo: [NSLocalizedDescriptionKey: "fragCount outside 1..\(BleRecordConstants.maxFragments)"])
        }
        guard fragIndex >= 0, fragIndex < fragCount else {
            throw NSError(domain: "BleRecordCodec", code: 3, userInfo: [NSLocalizedDescriptionKey: "fragIndex outside 0..<\(fragCount)"])
        }

        let stride = (totalLen + fragCount - 1) / fragCount
        let start = fragIndex * stride
        guard start < totalLen else {
            throw NSError(domain: "BleRecordCodec", code: 4, userInfo: [NSLocalizedDescriptionKey: "start >= totalLen"])
        }
        let end = min(start + stride, totalLen)
        return (start, end)
    }

    public static func encodeHeader(
        recordType: BleRecordType,
        recordSeq: UInt8,
        fragIndex: UInt8,
        fragCount: UInt8,
        totalLen: UInt16,
        magic: UInt8 = BleRecordConstants.magic
    ) throws -> Data {
        guard fragCount >= 1, fragCount <= BleRecordConstants.maxFragments else {
            throw NSError(domain: "BleRecordCodec", code: 10, userInfo: [NSLocalizedDescriptionKey: "fragCount out of bounds"])
        }
        guard fragIndex < fragCount else {
            throw NSError(domain: "BleRecordCodec", code: 11, userInfo: [NSLocalizedDescriptionKey: "fragIndex >= fragCount"])
        }
        guard totalLen <= BleRecordConstants.maxRecord else {
            throw NSError(domain: "BleRecordCodec", code: 12, userInfo: [NSLocalizedDescriptionKey: "totalLen > maxRecord"])
        }

        let b0 = magic
        let b1 = recordType.rawValue
        let b2 = recordSeq
        let b3 = fragIndex
        let b4 = fragCount
        let b5 = UInt8((totalLen >> 8) & 0xFF)
        let b6 = UInt8(totalLen & 0xFF)
        let b7 = computeHeaderCheck(b0, b1, b2, b3, b4, b5, b6)

        return Data([b0, b1, b2, b3, b4, b5, b6, b7])
    }

    public static func decodeHeader(_ data: Data) -> BleRecordHeader? {
        guard data.count >= BleRecordConstants.headerBytes else { return nil }

        let bytes = [UInt8](data.prefix(BleRecordConstants.headerBytes))
        let b0 = bytes[0]
        guard b0 == BleRecordConstants.magic else { return nil }

        guard let recordType = BleRecordType(rawValue: bytes[1]) else { return nil }

        let b2 = bytes[2]
        let b3 = bytes[3]
        let b4 = bytes[4]
        let totalLen = (UInt16(bytes[5]) << 8) | UInt16(bytes[6])
        let b7 = bytes[7]

        let expectedChk = computeHeaderCheck(b0, bytes[1], b2, b3, b4, bytes[5], bytes[6])
        guard b7 == expectedChk else { return nil }

        guard b4 >= 1, b4 <= BleRecordConstants.maxFragments else { return nil }
        guard b3 < b4 else { return nil }
        guard totalLen <= BleRecordConstants.maxRecord else { return nil }

        return BleRecordHeader(
            magic: b0,
            recordType: recordType,
            recordSeq: b2,
            fragIndex: b3,
            fragCount: b4,
            totalLen: totalLen,
            headerCheck: b7
        )
    }

    public static func decodeFragment(_ data: Data) -> BleRecordFragment? {
        guard data.count >= BleRecordConstants.headerBytes else { return nil }

        guard let header = decodeHeader(data) else { return nil }

        guard let (start, end) = try? canonicalFragmentBounds(
            totalLen: Int(header.totalLen),
            fragCount: Int(header.fragCount),
            fragIndex: Int(header.fragIndex)
        ) else {
            return nil
        }

        let expectedPayloadLen = end - start
        let actualPayloadLen = data.count - BleRecordConstants.headerBytes
        guard actualPayloadLen == expectedPayloadLen else { return nil }

        let payload = Data(data.dropFirst(BleRecordConstants.headerBytes))
        return BleRecordFragment(header: header, payload: payload)
    }
}

public enum BleRecordFragmenter {

    public static func fragment(
        recordType: BleRecordType,
        recordSeq: UInt8,
        payload: Data,
        maxAttValueLength: Int
    ) throws -> [Data] {
        let capacity = maxAttValueLength - BleRecordConstants.headerBytes
        guard capacity >= 1 else {
            throw NSError(domain: "BleRecordFragmenter", code: 1, userInfo: [NSLocalizedDescriptionKey: "capacity < 1"])
        }

        let totalLen = payload.count
        guard totalLen <= BleRecordConstants.maxRecord else {
            throw NSError(domain: "BleRecordFragmenter", code: 2, userInfo: [NSLocalizedDescriptionKey: "payload size > maxRecord"])
        }

        if totalLen == 0 {
            let hdr = try BleRecordCodec.encodeHeader(
                recordType: recordType,
                recordSeq: recordSeq,
                fragIndex: 0,
                fragCount: 1,
                totalLen: 0
            )
            return [hdr]
        }

        let fragCount = (totalLen + capacity - 1) / capacity
        guard fragCount <= BleRecordConstants.maxFragments else {
            throw NSError(domain: "BleRecordFragmenter", code: 3, userInfo: [NSLocalizedDescriptionKey: "fragCount > maxFragments"])
        }

        let stride = (totalLen + fragCount - 1) / fragCount
        var fragments: [Data] = []
        fragments.reserveCapacity(fragCount)

        let payloadBytes = [UInt8](payload)
        for i in 0..<fragCount {
            let start = i * stride
            let end = min(start + stride, totalLen)
            let fragPayload = Data(payloadBytes[start..<end])
            let hdr = try BleRecordCodec.encodeHeader(
                recordType: recordType,
                recordSeq: recordSeq,
                fragIndex: UInt8(i),
                fragCount: UInt8(fragCount),
                totalLen: UInt16(totalLen)
            )
            var fullFrag = hdr
            fullFrag.append(fragPayload)
            fragments.append(fullFrag)
        }

        return fragments
    }
}

public final class BleRecordReassembler: @unchecked Sendable {

    private final class InFlightAssembly {
        let recordType: BleRecordType
        let totalLen: Int
        let fragCount: Int
        let stride: Int
        var receivedIndices: Set<Int>
        var buffer: [UInt8]
        let createdTime: TimeInterval
        var lastActivityTime: TimeInterval

        init(
            recordType: BleRecordType,
            totalLen: Int,
            fragCount: Int,
            stride: Int,
            createdTime: TimeInterval
        ) {
            self.recordType = recordType
            self.totalLen = totalLen
            self.fragCount = fragCount
            self.stride = stride
            self.receivedIndices = []
            self.buffer = [UInt8](repeating: 0, count: totalLen)
            self.createdTime = createdTime
            self.lastActivityTime = createdTime
        }
    }

    private let timeProvider: () -> TimeInterval
    private var inFlight: [UInt8: InFlightAssembly] = [:]
    private var completedFingerprints: [UInt8: String] = [:]
    private let lock = NSLock()

    public init(timeProvider: @escaping () -> TimeInterval = { Date().timeIntervalSince1970 }) {
        self.timeProvider = timeProvider
    }

    private func evictExpired(now: TimeInterval) {
        var expired: [UInt8] = []
        for (seq, asm) in inFlight {
            if (now - asm.lastActivityTime) > BleRecordConstants.reassemblyTimeoutSeconds {
                expired.append(seq)
            }
        }
        for seq in expired {
            inFlight.removeValue(forKey: seq)
        }
    }

    private func computeFingerprint(recordType: BleRecordType, seq: UInt8, payload: Data) -> String {
        var hasher = SHA256()
        hasher.update(data: Data([recordType.rawValue, seq]))
        hasher.update(data: payload)
        let digest = hasher.finalize()
        return digest.map { String(format: "%02x", $0) }.joined()
    }

    public func receiveFragmentBytes(_ data: Data) -> BleReassembledRecord? {
        guard let frag = BleRecordCodec.decodeFragment(data) else { return nil }
        return receiveFragment(frag)
    }

    public func receiveFragment(_ frag: BleRecordFragment) -> BleReassembledRecord? {
        lock.lock()
        defer { lock.unlock() }

        let now = timeProvider()
        evictExpired(now: now)

        let hdr = frag.header
        let seq = hdr.recordSeq

        let totalLen = Int(hdr.totalLen)
        let fragCount = Int(hdr.fragCount)
        let fragIndex = Int(hdr.fragIndex)

        let asm: InFlightAssembly
        if let existing = inFlight[seq] {
            asm = existing
        } else {
            // Check capacity
            guard inFlight.count < BleRecordConstants.maxConcurrent else {
                return nil // Reject 5th concurrent assembly
            }

            let stride = totalLen > 0 ? (totalLen + fragCount - 1) / fragCount : 0
            let newAsm = InFlightAssembly(
                recordType: hdr.recordType,
                totalLen: totalLen,
                fragCount: fragCount,
                stride: stride,
                createdTime: now
            )
            inFlight[seq] = newAsm
            asm = newAsm
        }

        // Validate metadata consistency
        guard asm.recordType == hdr.recordType,
              asm.totalLen == totalLen,
              asm.fragCount == fragCount else {
            inFlight.removeValue(forKey: seq)
            return nil
        }

        asm.lastActivityTime = now

        guard let (start, end) = try? BleRecordCodec.canonicalFragmentBounds(
            totalLen: totalLen,
            fragCount: fragCount,
            fragIndex: fragIndex
        ) else {
            inFlight.removeValue(forKey: seq)
            return nil
        }

        let fragBytes = [UInt8](frag.payload)

        // Check duplicate fragment index
        if asm.receivedIndices.contains(fragIndex) {
            let existingPayload = Array(asm.buffer[start..<end])
            if existingPayload == fragBytes {
                // Idempotent duplicate
                return nil
            } else {
                // Conflicting duplicate: invalidate and fail closed
                inFlight.removeValue(forKey: seq)
                return nil
            }
        }

        // Store fragment payload
        for (offset, byte) in fragBytes.enumerated() {
            asm.buffer[start + offset] = byte
        }
        asm.receivedIndices.insert(fragIndex)

        // Check if complete
        if asm.receivedIndices.count == asm.fragCount {
            let completePayload = Data(asm.buffer)
            inFlight.removeValue(forKey: seq)

            let fp = computeFingerprint(recordType: hdr.recordType, seq: seq, payload: completePayload)
            if completedFingerprints[seq] == fp {
                // Duplicate completion
                return nil
            }

            completedFingerprints[seq] = fp
            return BleReassembledRecord(
                recordType: hdr.recordType,
                recordSeq: seq,
                payload: completePayload
            )
        }

        return nil
    }

    public func reset() {
        lock.lock()
        defer { lock.unlock() }
        inFlight.removeAll()
        completedFingerprints.removeAll()
    }
}

File: android/mesh/src/main/java/io/godstone/mesh/transport/BleRecord.kt
package io.godstone.mesh.transport

import java.security.MessageDigest

/**
 * ADR-002 Canonical BLE Record Layer Types & Constants (Phase C8.4C).
 */
enum class BleRecordType(val typeCode: Byte) {
    HS1(0x11.toByte()),
    HS2(0x12.toByte()),
    HS3(0x14.toByte()),
    DATA(0x18.toByte()),
    CLOSE(0x21.toByte());

    companion object {
        fun fromByte(b: Byte): BleRecordType? = entries.firstOrNull { it.typeCode == b }
    }
}

object BleRecordConstants {
    const val MAGIC: Byte = 0x47.toByte()
    const val HEADER_BYTES: Int = 8
    const val MAX_RECORD: Int = 16384
    const val MAX_FRAGMENTS: Int = 64
    const val MAX_CONCURRENT: Int = 4
    const val REASSEMBLY_TIMEOUT_SECONDS: Long = 30L
}

data class BleRecordHeader(
    val magic: Byte,
    val recordType: BleRecordType,
    val recordSeq: Int,
    val fragIndex: Int,
    val fragCount: Int,
    val totalLen: Int,
    val headerCheck: Byte,
)

class BleRecordFragment(
    val header: BleRecordHeader,
    payload: ByteArray,
) {
    val payload: ByteArray = payload.copyOf()

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is BleRecordFragment) return false
        return header == other.header && payload.contentEquals(other.payload)
    }

    override fun hashCode(): Int = 31 * header.hashCode() + payload.contentHashCode()
}

class BleReassembledRecord(
    val recordType: BleRecordType,
    val recordSeq: Int,
    payload: ByteArray,
) {
    val payload: ByteArray = payload.copyOf()

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is BleReassembledRecord) return false
        return recordType == other.recordType && recordSeq == other.recordSeq && payload.contentEquals(other.payload)
    }

    override fun hashCode(): Int {
        var result = recordType.hashCode()
        result = 31 * result + recordSeq
        result = 31 * result + payload.contentHashCode()
        return result
    }
}

object BleRecordCodec {

    fun computeHeaderCheck(
        b0: Byte,
        b1: Byte,
        b2: Byte,
        b3: Byte,
        b4: Byte,
        b5: Byte,
        b6: Byte,
    ): Byte {
        return (b0.toInt() xor b1.toInt() xor b2.toInt() xor b3.toInt() xor b4.toInt() xor b5.toInt() xor b6.toInt()).toByte()
    }

    fun computeHeaderCheck(bytes7: ByteArray): Byte {
        require(bytes7.size == 7) { "computeHeaderCheck requires exactly 7 bytes, got ${bytes7.size}" }
        return computeHeaderCheck(
            bytes7[0],
            bytes7[1],
            bytes7[2],
            bytes7[3],
            bytes7[4],
            bytes7[5],
            bytes7[6],
        )
    }

    fun canonicalFragmentBounds(totalLen: Int, fragCount: Int, fragIndex: Int): Pair<Int, Int> {
        if (totalLen == 0) {
            require(fragCount == 1 && fragIndex == 0) {
                "totalLen=0 requires fragCount=1, fragIndex=0; got count=$fragCount, index=$fragIndex"
            }
            return Pair(0, 0)
        }
        require(fragCount in 1..BleRecordConstants.MAX_FRAGMENTS) {
            "fragCount $fragCount outside 1..${BleRecordConstants.MAX_FRAGMENTS}"
        }
        require(fragIndex in 0 until fragCount) {
            "fragIndex $fragIndex outside 0 until $fragCount"
        }
        val stride = (totalLen + fragCount - 1) / fragCount
        val start = fragIndex * stride
        require(start < totalLen) {
            "canonical start $start >= totalLen $totalLen"
        }
        val end = minOf(start + stride, totalLen)
        return Pair(start, end)
    }

    fun encodeHeader(
        recordType: BleRecordType,
        recordSeq: Int,
        fragIndex: Int,
        fragCount: Int,
        totalLen: Int,
        magic: Byte = BleRecordConstants.MAGIC,
    ): ByteArray {
        require(recordSeq in 0..255) { "recordSeq $recordSeq outside 0..255" }
        require(fragCount in 1..BleRecordConstants.MAX_FRAGMENTS) {
            "fragCount $fragCount outside 1..${BleRecordConstants.MAX_FRAGMENTS}"
        }
        require(fragIndex in 0 until fragCount) {
            "fragIndex $fragIndex outside 0 until $fragCount"
        }
        require(totalLen in 0..BleRecordConstants.MAX_RECORD) {
            "totalLen $totalLen outside 0..${BleRecordConstants.MAX_RECORD}"
        }

        val b0 = magic
        val b1 = recordType.typeCode
        val b2 = recordSeq.toByte()
        val b3 = fragIndex.toByte()
        val b4 = fragCount.toByte()
        val b5 = ((totalLen ushr 8) and 0xFF).toByte()
        val b6 = (totalLen and 0xFF).toByte()
        val b7 = computeHeaderCheck(b0, b1, b2, b3, b4, b5, b6)

        return byteArrayOf(b0, b1, b2, b3, b4, b5, b6, b7)
    }

    fun decodeHeader(bytes: ByteArray): BleRecordHeader? {
        if (bytes.size < BleRecordConstants.HEADER_BYTES) return null

        val magic = bytes[0]
        if (magic != BleRecordConstants.MAGIC) return null

        val recordType = BleRecordType.fromByte(bytes[1]) ?: return null
        val recordSeq = bytes[2].toInt() and 0xFF
        val fragIndex = bytes[3].toInt() and 0xFF
        val fragCount = bytes[4].toInt() and 0xFF
        val totalLen = ((bytes[5].toInt() and 0xFF) shl 8) or (bytes[6].toInt() and 0xFF)
        val chk = bytes[7]

        val expectedChk = computeHeaderCheck(
            bytes[0], bytes[1], bytes[2], bytes[3], bytes[4], bytes[5], bytes[6]
        )
        if (chk != expectedChk) return null

        if (fragCount !in 1..BleRecordConstants.MAX_FRAGMENTS) return null
        if (fragIndex !in 0 until fragCount) return null
        if (totalLen !in 0..BleRecordConstants.MAX_RECORD) return null

        return BleRecordHeader(
            magic = magic,
            recordType = recordType,
            recordSeq = recordSeq,
            fragIndex = fragIndex,
            fragCount = fragCount,
            totalLen = totalLen,
            headerCheck = chk,
        )
    }

    fun decodeFragment(bytes: ByteArray): BleRecordFragment? {
        if (bytes.size < BleRecordConstants.HEADER_BYTES) return null

        val header = decodeHeader(bytes) ?: return null

        val (start, end) = try {
            canonicalFragmentBounds(header.totalLen, header.fragCount, header.fragIndex)
        } catch (_: Exception) {
            return null
        }

        val expectedPayloadLen = end - start
        val actualPayloadLen = bytes.size - BleRecordConstants.HEADER_BYTES
        if (actualPayloadLen != expectedPayloadLen) return null

        val payload = bytes.copyOfRange(BleRecordConstants.HEADER_BYTES, bytes.size)
        return BleRecordFragment(header = header, payload = payload)
    }
}

object BleRecordFragmenter {

    fun fragment(
        recordType: BleRecordType,
        recordSeq: Int,
        payload: ByteArray,
        maxAttValueLength: Int,
    ): List<ByteArray> {
        val capacity = maxAttValueLength - BleRecordConstants.HEADER_BYTES
        require(capacity >= 1) {
            "maxAttValueLength $maxAttValueLength must be >= ${BleRecordConstants.HEADER_BYTES + 1}"
        }

        val totalLen = payload.size
        require(totalLen <= BleRecordConstants.MAX_RECORD) {
            "payload size $totalLen exceeds MAX_RECORD ${BleRecordConstants.MAX_RECORD}"
        }

        if (totalLen == 0) {
            val hdr = BleRecordCodec.encodeHeader(recordType, recordSeq, 0, 1, 0)
            return listOf(hdr)
        }

        val fragCount = (totalLen + capacity - 1) / capacity
        require(fragCount <= BleRecordConstants.MAX_FRAGMENTS) {
            "fragCount $fragCount exceeds MAX_FRAGMENTS ${BleRecordConstants.MAX_FRAGMENTS}"
        }

        val stride = (totalLen + fragCount - 1) / fragCount
        val fragments = ArrayList<ByteArray>(fragCount)

        for (i in 0 until fragCount) {
            val start = i * stride
            val end = minOf(start + stride, totalLen)
            val fragPayload = payload.copyOfRange(start, end)
            val hdr = BleRecordCodec.encodeHeader(recordType, recordSeq, i, fragCount, totalLen)
            val fullFrag = ByteArray(BleRecordConstants.HEADER_BYTES + fragPayload.size)
            System.arraycopy(hdr, 0, fullFrag, 0, BleRecordConstants.HEADER_BYTES)
            System.arraycopy(fragPayload, 0, fullFrag, BleRecordConstants.HEADER_BYTES, fragPayload.size)
            fragments.add(fullFrag)
        }

        return fragments
    }
}

class BleRecordReassembler(
    private val clock: () -> Long = { System.currentTimeMillis() / 1000L },
) {
    private class InFlightAssembly(
        val recordType: BleRecordType,
        val totalLen: Int,
        val fragCount: Int,
        val stride: Int,
        val receivedIndices: MutableSet<Int>,
        val buffer: ByteArray,
        val createdTimeSec: Long,
        var lastActivityTimeSec: Long,
    )

    private val inFlight = HashMap<Int, InFlightAssembly>()
    private val completedFingerprints = HashMap<Int, String>()

    private fun evictExpired(nowSec: Long) {
        val expiredSeqs = inFlight.filter { (_, asm) ->
            (nowSec - asm.lastActivityTimeSec) > BleRecordConstants.REASSEMBLY_TIMEOUT_SECONDS
        }.keys.toList()

        for (seq in expiredSeqs) {
            inFlight.remove(seq)
        }
    }

    private fun computeFingerprint(recordType: BleRecordType, seq: Int, payload: ByteArray): String {
        val md = MessageDigest.getInstance("SHA-256")
        md.update(recordType.typeCode)
        md.update(seq.toByte())
        md.update(payload)
        return md.digest().joinToString("") { "%02x".format(it) }
    }

    fun receiveFragmentBytes(rawBytes: ByteArray): BleReassembledRecord? {
        val frag = BleRecordCodec.decodeFragment(rawBytes) ?: return null
        return receiveFragment(frag)
    }

    fun receiveFragment(frag: BleRecordFragment): BleReassembledRecord? {
        val nowSec = clock()
        evictExpired(nowSec)

        val hdr = frag.header
        val seq = hdr.recordSeq

        var asm = inFlight[seq]
        if (asm == null) {
            // Check capacity
            if (inFlight.size >= BleRecordConstants.MAX_CONCURRENT) {
                return null // Reject 5th concurrent assembly
            }

            val stride = if (hdr.totalLen > 0) (hdr.totalLen + hdr.fragCount - 1) / hdr.fragCount else 0
            asm = InFlightAssembly(
                recordType = hdr.recordType,
                totalLen = hdr.totalLen,
                fragCount = hdr.fragCount,
                stride = stride,
                receivedIndices = HashSet(),
                buffer = ByteArray(hdr.totalLen),
                createdTimeSec = nowSec,
                lastActivityTimeSec = nowSec,
            )
            inFlight[seq] = asm
        }

        // Validate metadata consistency
        if (asm.recordType != hdr.recordType || asm.totalLen != hdr.totalLen || asm.fragCount != hdr.fragCount) {
            // Conflicting metadata: invalidate and fail closed
            inFlight.remove(seq)
            return null
        }

        asm.lastActivityTimeSec = nowSec

        val (start, end) = try {
            BleRecordCodec.canonicalFragmentBounds(hdr.totalLen, hdr.fragCount, hdr.fragIndex)
        } catch (_: Exception) {
            inFlight.remove(seq)
            return null
        }

        // Check duplicate fragment index
        if (hdr.fragIndex in asm.receivedIndices) {
            val existingPayload = asm.buffer.copyOfRange(start, end)
            return if (existingPayload.contentEquals(frag.payload)) {
                // Idempotent duplicate: ignore
                null
            } else {
                // Conflicting duplicate: invalidate and fail closed
                inFlight.remove(seq)
                null
            }
        }

        // Store fragment payload
        System.arraycopy(frag.payload, 0, asm.buffer, start, end - start)
        asm.receivedIndices.add(hdr.fragIndex)

        // Check if complete
        if (asm.receivedIndices.size == asm.fragCount) {
            val completePayload = asm.buffer.copyOf()
            inFlight.remove(seq)

            val fp = computeFingerprint(hdr.recordType, seq, completePayload)
            if (completedFingerprints[seq] == fp) {
                // Duplicate retransmission of already completed record
                return null
            }

            completedFingerprints[seq] = fp
            return BleReassembledRecord(
                recordType = hdr.recordType,
                recordSeq = seq,
                payload = completePayload,
            )
        }

        return null
    }

    fun reset() {
        inFlight.clear()
        completedFingerprints.clear()
    }
}
